addons/mail/models/mail_thread.py

```diff
@@ -3652,30 +3652,12 @@ def _notify_by_email_get_final_mail_values(self, recipient_ids, mail_values,
             final_mail_values.update(additional_values)
         return final_mail_values
 
-    def _notify_thread_by_web_push(self, message, recipients_data, msg_vals=False, **kwargs):
-        """ Method to send cloud notifications for every mention of a partner
-        and every direct message. We have to take into account the risk of
-        duplicated notifications in case of a mention in a channel of `chat` type.
-
-        :param message: ``mail.message`` record to notify;
-        :param recipients_data: list of recipients information (based on res.partner
-          records), formatted like
-            [{'active': partner.active;
-              'id': id of the res.partner being recipient to notify;
-              'groups': res.group IDs if linked to a user;
-              'notif': 'inbox', 'email', 'sms' (SMS App);
-              'share': partner.partner_share;
-              'type': 'customer', 'portal', 'user;'
-             }, {...}].
-          See ``MailThread._notify_get_recipients``;
-        :param msg_vals: dictionary of values used to create the message. If given it
-          may be used to access values related to ``message`` without accessing it
-          directly. It lessens query count in some optimized use cases by avoiding
-          access message content in db;
+    def _push_web_notification(self, get_payload, partner_ids):
```

Comment:
Would keep definition of _notify_thread_by_web_push first (both to have main method first which is used in this file (at least generally), and to lessen diff as we will normally avoid the diff of docstring and stuff like that)

Reply:
Thanks for the review, I applied the other comments, 

For this one the reason I pass a function instead of evaluting the payload upstream is that there are extra querries when forming the payload that are avoided if it is evaluated only when needed (for example when there is no device, vapid keys or partners)

Reply:
This could maybe be extracted, and we skip the call in that case ? I really think giving function pointers is a mess to understand code, and actually it looks simpler to me to 1/ check if we have to do push notifications and 2/ do it. Moreover if we do it, I would do the check on partner_ids first (as there is no cost), then config parameter first (as those are cached), then devices (to do the search only if we are sure we have to do it).